Enforce POST requests must have json content type, or route override

dpsutton opened the pull request:
Want to ensure that we reject post requests to our backend if they have a type different from application/json. This was attempted in the past in https://github.com/metabase/metabase/pull/9547/files and then undone in https://github.com/metabase/metabase/pull/9559/files.

The issue was that some of our query endpoints and download routes take form params, so we can't just do a blanket "if POST, require application/json".

But the routes like this are few and far between. The solution:

```clojure
(api/defendpoint ^:streaming ^{:content-types #{:content/json :content/form}} POST
  "/:card-id/query/:export-format"
  ...)
```

This specifies that this route for `POST api/card/<card-id>/query/csv` can accept _either_ form or json headers. There are a few tests that might still send a json request so we can leave it for now.

How it works:

Compojure depends on a library called clout. This is the actual routing bit in a protocol

```clojure
(defprotocol Route
  (route-matches [route request]
    "If the route matches the supplied request, the matched keywords are
    returned as a map. Otherwise, nil is returned."))
```

So for posts, we allow it to construct that implementation, but then we wrap it. The wrapper checks if it matches, and then if so, checks if the content-type headers are appropriate.

The matching is based on regexes. Matches a grep-friendly keyword (`:content/json`, `:content/form`, rather than trying to find instances of `:json` or `:form` in the codebase).

```clojure
(def ^:private content-type->regex
  "A map of content type to a regex that should match that content type."
  {:content/json #"^application/(.+\+)?json"
   :content/form #"(^application/x-www-form-urlencoded)|(multipart/form-data)|(application/form)"})
```

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/26520)
<!-- Reviewable:end -->

dpsutton committed:
Enforce POST requests must have json content type, or route override

Want to ensure that we reject post requests to our backend if they have
a type different from application/json. This was attempted in the past
in https://github.com/metabase/metabase/pull/9547/files and then undone
in https://github.com/metabase/metabase/pull/9559/files.

The issue was that some of our query endpoints and download routes take
form params, so we can't just do a blanket "if POST, require
application/json".

But the routes like this are few and far between. The solution:

```clojure
(api/defendpoint ^:streaming ^{:content-types #{:content/json :content/form}} POST
  "/:card-id/query/:export-format"
  ...)
```

This specifies that this route for `POST api/card/<card-id>/query/csv`
can accept _either_ form or json headers. There are a few tests that
might still send a json request so we can leave it for now.

How it works:

Compojure depends on a library called clout. This is the actual routing
bit in a protocol

```clojure
(defprotocol Route
  (route-matches [route request]
    "If the route matches the supplied request, the matched keywords are
    returned as a map. Otherwise, nil is returned."))
```

So for posts, we allow it to construct that implementation, but then we
wrap it. The wrapper checks if it matches, and then if so, checks if the
content-type headers are appropriate.

The matching is based on regexes. Matches a grep-friendly
keyword (`:content/json`, `:content/form`, rather than trying to find
instances of `:json` or `:form` in the codebase).

```clojure
(def ^:private content-type->regex
  "A map of content type to a regex that should match that content type."
  {:content/json #"^application/(.+\+)?json"
   :content/form #"(^application/x-www-form-urlencoded)|(multipart/form-data)|(application/form)"})
```

dpsutton committed:
include test for route with regex

dpsutton committed:
sso takes form params

dpsutton committed:
testing endpoint /restore/:name doesn't care and gets wildcard

dpsutton committed:
handle nil content-type

dpsutton committed:
Fix rest of testing api endpoints

dpsutton committed:
cypress tests hit the api/database/:id/sync_schema

qnkhuat commented on src/metabase/api/common.clj:
```diff
@@ -255,6 +256,51 @@
+  "A map of content type to a regex that should match that content type."
+  {:content/json #"^application/(.+\+)?json"
+   :content/form #"(^application/x-www-form-urlencoded)|(multipart/form-data)|(application/form)"
+   :content/*    (constantly true)})
```
```suggestion
   :content/*    #".*"})
```

nit: we can do this so the `content-type-matches` don't have to deal with the fact that `value` of `content-type->regex` could be a func or a regex?

dpsutton replied:
I was originally going to do this. The problem is that often requests will not have a content type so it would be a regex against a nil string. Changing to a function that just constantly returns true is exactly what we want, regardless of the content-type or lack of it.

dpsutton replied:
if we really want uniformity, perhaps the regexes could just be partial'd so that all of the matchers are just functions.

camsaul commented on enterprise/backend/src/metabase_enterprise/sso/api/sso.clj:
```diff
@@ -43,7 +43,7 @@
                  :additionalData data}))})
 
-(api/defendpoint POST "/"
+(api/defendpoint ^{:content-types #{:content/form}} POST "/"
```
Are you sure this should only be `:content/form`? This is the endpoint we get in the SSO response from the IdP... not sure what `Content-Type` they're sending us but I would have guessed there would be a good chance it could be JSON

dpsutton replied:
i can add both. Or perhaps this could be `:content/*`. I have no hope that this will hit the release and should bake in master for a while. I'm also considering an env var to turn it off in case it proves unworkable in some cases.

camsaul replied:
I think `:content/*` makes sense here. I don't think we need an env var for this -- we really don't support anything besides `Content-Type: application/json` for anything except for a handful of endpoints. An env var to disable the error wouldn't change that fact. I like keeping it required and giving it time in `master` to bake

dpsutton replied:
The issue is happening in CI. Lots of things don't send a content type at all. I'm making a change that if there are no bindings for the route other than route-provided bindings (card-id from the route itself) then it does not enforce the content type. But its possible lots of api usage (or maybe our browser client?) won't set this. Seemed safer to be able to disable it if it proved buggy at first.

camsaul commented on src/metabase/api/card.clj:
```diff
@@ -784,7 +784,8 @@ saved later when it is ready."
    :middleware   {:process-viz-settings? false}))
 
-(api/defendpoint ^:streaming POST "/:card-id/query/:export-format"
+(api/defendpoint ^:streaming ^{:content-types #{:content/json :content/form}} POST
```
Is there ever going to be something in this set that doesn't start with `:content/`? Seems like noise

dpsutton replied:
My plan is that they would all be `:content/<something>`. I think this should make it easier to find all usages of the keywords. If we revert to un-namespaced, finding where this stuff works by looking for `:json` would be quite difficult. I hope this doesn't appear as noise because i consider it signal. If not, we could move to longer un-namespaced keywords but they would probably be similar to `:content-json` at which point i think we aren't gaining anything.

camsaul commented on src/metabase/api/common.clj:
```diff
@@ -255,6 +256,51 @@
+  [route method content-types]
+  (if (= (str/lower-case (name method)) "post")
+    (do
+      (when (seq content-types)
+        (run! (fn [content-type]
+                (or (content-type->regex content-type)
+                    (throw (ex-info (trs "Unrecognized content type: {0}\n Valid: {1}"
+                                         content-type
+                                         (keys content-type->regex))
+                                    {}))))
+              content-types))
```
Could you do this validation as part of the `s/fdef` for `defendpoint`? See if the method is the symbol `POST`, and if it is validate the metadata or something. Idk it might be quicker feedback than doing it this way

camsaul commented on src/metabase/api/common.clj:
```diff
@@ -263,7 +309,11 @@
                     :doc          docstr
                     :is-endpoint? true)
-     (~(symbol "compojure.core" (name method)) ~route ~args
+     (~(symbol "compojure.core" (name method))
```
Seeing how we qualify the method in the macro anyway maybe we should update the kondo linter so it stops complaining if you don't explicitly `:refer` `GET` or `POST` or whatever

dpsutton replied:
Yeah that struck me. The macro just cares that they are the symbols `POST` or `GET`, etc. and not that they are actually vars.

camsaul commented on src/metabase/api/database.clj:
```diff
@@ -881,7 +881,7 @@
 
 ;; Should somehow trigger sync-database/sync-database!
-(api/defendpoint POST "/:id/sync_schema"
+(api/defendpoint ^{:content-types #{:content/*}} POST "/:id/sync_schema"
```
For POST endpoints that don't expect a body does it make sense to validate the `Content-Type`? Maybe what we should actually be validating is that the body is empty or something

camsaul replied:
Or maybe only validate Content-Type if body is present... it seems a little dangerous that we have to remember to put `^{:content-types #{:content/*}}` on every endpoint that doesn't expect a body now or risk breaking things for people

dpsutton replied:
Yeah i thought about that. Some issues: we've have to manually annotate which ones don't read the request which could get wild.
As for reading the request, I'm not sure that works. If it's json, we eagerly parse that in the middleware. For others, we kinda want the input stream to be read by the handler.

If this proves onerous, perhaps we could switch and annotate which ones we want to enforce have a content type rather than adding exceptions for ones we don't care about.

dpsutton replied:
This was an awesome thought actually. I have the args, and can get the args pulled from the route. If all args come from the route, then we can just ignore the content-type completely. Good thinking.

camsaul reviewed (approved):
Well overall this looks good to me, definitely an improvement over existing stuff. I left a few notes but those are more me thinking aloud than anything else

Question: why stop at only validating `POST`... why not validate `PUT` as well? Not as important for CSRF protection purposes but it's still nice to tell people to stop trying to send us XML or other wacky stuff to knock it off, rather than let them sit around and scratch their heads wondering why it doesn't work

dpsutton committed:
Only assert on content type for routes which use the body

macro has the endpoint args and can get the route args. If those are
identical, then nothing comes from the body and we can ignore
content-type.

  Eg: route:       "/:card-id/query/:export-format"
      args vector: [card-id export-format]
does not require a check of content type because its only args are from
the route itself.

      route:       "/:card-id/query/:export-format"
      args vector: [card-id export-format :as {{:keys [parameters]} :params}]
  does require a check because it pulls parameters from the request."

dpsutton committed:
Simplify the route handling

start using the content-length (if available) to see if the body is
empty. empty-bodies means we don't care about content-type and so all of
the testing apis and posting to database sync will be fine.

dpsutton committed:
linter cleanup / typehints

dpsutton committed:
mark dataset endpoint as content/form

dpsutton committed:
safely read body on request

dpsutton committed:
allow json and form on POST api/dataset

we actually expect application/form (or some variant x-www-formurlencode
or whatever). But lots of tests seem to want to just send a payload so
we'll just allow json here as well.

dpsutton committed:
Make invalid content-type a 400 response, add form to xlsx/csv download route

dpsutton committed:
Require content-types on PUTs

camsaul reviewed (approved):
> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/metabase/metabase/26520)* status: all files reviewed, 3 unresolved discussions (waiting on @dpsutton and @qnkhuat)

___
*[`test/metabase/api/common_test.clj` line 117 at r1](https://reviewable.io/reviews/metabase/metabase/26520#-NRjMeny1dlWNagQ74jS:-NRjMeny1dlWNagQ74jT:b-m9vs78) ([raw file](https://github.com/metabase/metabase/blob/9f1aba953adcd705d19d660dfb51132044f6f4a0/test/metabase/api/common_test.clj#L117)):*
> ```Clojure
> 
> 
> (deftest matching-route-test
> ```

Can this test be parallel?



_Code quote (i):_
```Clojure
deftest matching-route-test
```



_Code snippet (ii):_
```Clojure
deftest ^:parallel matching-route-test
```
___
*[`test/metabase/api/common_test.clj` line 227 at r1](https://reviewable.io/reviews/metabase/metabase/26520#-NRjMvQMDgkpd3aQVdpM:-NRjMvQMDgkpd3aQVdpN:b-6n8e3d) ([raw file](https://github.com/metabase/metabase/blob/9f1aba953adcd705d19d660dfb51132044f6f4a0/test/metabase/api/common_test.clj#L227)):*
> ```Clojure
> (api/define-routes)
> 
> (deftest post-and-put-routing-test
> ```

Likewise, it seems like there's no reason this test can't be parallel



_Code quote (i):_
```Clojure
(deftest post-and-put-routing-test
```



_Code snippet (ii):_
```Clojure
(deftest ^:parallel post-and-put-routing-test
```


<!-- Sent from Reviewable.io -->
